msos/StackWalker.cs

```diff
@@ -0,0 +1,372 @@
+                        }
+                    }
+                    finally
+                    {
+                        if (pointer != default(IntPtr))
+                        {
+                            Marshal.FreeHGlobal(pointer);
+                        }
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Gets Handle Object name using NtQueryObject NtDll function
+        /// Doc: https://msdn.microsoft.com/en-us/library/bb432383(v=vs.85).aspx
+        /// </summary>
+        /// <param name="handle"></param>
+        /// <returns></returns>
+        public static unsafe string GetHandleObjectName(IntPtr handle, uint pid)
+        {
+            string result = null;
+
+            IntPtr duplicatedHandle = default(IntPtr);
+
+            if (Duplicate(handle, pid, out duplicatedHandle))
+            {
+                int length;
+
+                NtStatus stat = NtQueryObject(duplicatedHandle,
+                    OBJECT_INFORMATION_CLASS.ObjectNameInformation, IntPtr.Zero, 0, out length);
+
+                if (stat != NtStatus.InvalidHandle)
+                {
+                    IntPtr pointer = default(IntPtr);
+                    try
+                    {
+                        pointer = Marshal.AllocHGlobal((int)length);
+
+                        NtStatus status = NtQueryObject(duplicatedHandle,
+                               OBJECT_INFORMATION_CLASS.ObjectNameInformation, pointer, length, out length);
+
+                        if (status == NtStatus.Success)
+                        {
+
+                            OBJECT_NAME_INFORMATION info = (OBJECT_NAME_INFORMATION)Marshal.PtrToStructure(pointer, typeof(OBJECT_NAME_INFORMATION));
+                            result = info.Name.ToString();
+                        }
+                    }
+                    finally
+                    {
+                        if (pointer != default(IntPtr))
+                        {
+                            Marshal.FreeHGlobal(pointer);
+                        }
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Perform handle duplication
+        /// </summary>
+        /// <param name="handle">Handle which needed to be duplicated</param>
+        /// <param name="pid">Process PID (handle owner)</param>
+        /// <param name="result">Duplicated Handle</param>
+        /// <returns></returns>
+        private static bool Duplicate(IntPtr handle, uint pid, out IntPtr result)
+        {
+            result = default(IntPtr);
+
+            var processHandle = OpenProcess(ProcessAccessFlags.All, true, pid);
+
+            var process = GetCurrentProcess();
+            var options = DuplicateOptions.DUPLICATE_SAME_ACCESS;
+
+            return DuplicateHandle(processHandle, handle, process, out result, 0, false, options);
```

Comment:
Just want to make sure because I haven't noticed it elsewhere -- are you closing the handle afterwards? If not, it's a leak.